gazelle/sass/sass.go

```diff
@@ -0,0 +1,179 @@
+		},
+	}
+}
+
+// Loads returns .bzl files and symbols they define. Every rule generated by
+// GenerateRules, now or in the past, should be loadable from one of these
+// files.
+func (s *sasslang) Loads() []rule.LoadInfo {
+	return []rule.LoadInfo{
+		{
+			Name:    "@io_bazel_rules_sass//:defs.bzl",
+			Symbols: []string{"sass_binary", "sass_library"},
+		},
+	}
+}
+
+// GenerateRules extracts build metadata from source files in a directory.
+// GenerateRules is called in each directory where an update is requested
+// in depth-first post-order.
+//
+// args contains the arguments for GenerateRules. This is passed as a
+// struct to avoid breaking implementations in the future when new
+// fields are added.
+//
+// empty is a list of empty rules that may be deleted after merge.
+//
+// gen is a list of generated rules that may be updated or added.
+//
+// Any non-fatal errors this function encounters should be logged using
+// log.Print.
+func (s *sasslang) GenerateRules(args language.GenerateArgs) language.GenerateResult {
+	// base is the last part of the path for this element. For example:
+	// "hello_world" => "hello_world"
+	// "foo/bar" => "bar"
+	base := path.Base(args.Rel)
+	if base == "." {
+		//args.Rel will return an empty string if you're in the root of the repo.
+		//This will then be translated into "." by path.Base which is not a valid
+		//name for a target. Therefore we will use the name of "root" just to have
+		//something that is valid. If the user doesn't want the target to be named
+		// `//:root`, then they can rename it and on the next generation it will
+		// persist the user supplied name.
+		base = "root"
+	}
+
+	rules := []*rule.Rule{}
+	imports := []interface{}{}
+
+	var normalFiles []string
+	for _, f := range append(args.RegularFiles, args.GenFiles...) {
+		// Only generate Sass entries for sass files (.scss/.sass)
+		if !strings.HasSuffix(f, ".sass") && !strings.HasSuffix(f, ".scss") {
+			continue
+		}
+
+		fileInfo, err := sassFileInfo(args.Dir, f)
+		if err != nil {
+			panic(fmt.Sprintf("Fatal error"))
+		}
+
+		for err := range fileInfo.Errors {
+			log.Printf("Error parsing %s: %s\n", fileInfo.Name, err)
+		}
+
+		imports = append(imports, fileInfo.Imports)
+
+		// The primary entrypoint on Sass is a main.scss file.
+		if f == "main.scss" {
+			rule := rule.NewRule("sass_binary", base)
+
+			rule.SetAttr("src", "main.scss")
+			rules = append(rules, rule)
+		} else if strings.HasPrefix(path.Base(f), "_") {
+			// Libraries in Sass have filenames that start with _.
+			// For each file in the dir with a leading "_" create a new sass_library
+			rule := rule.NewRule("sass_library", base[1:len(base)-5])
+
+			rule.SetAttr("srcs", []string{base})
+			rule.SetPrivateAttr(config.GazelleImportsKey, fileInfo.Imports)
+
+			// These rules should always be public
+			rule.SetAttr("visibility", []string{"//visibility:public"})
+
+			rules = append(rules, rule)
+		} else {
+			normalFiles = append(normalFiles, f)
```

Comment:
Won't this make the imports array and the rules array not have the same length (or be associated by index?)
[GenerateResult](https://pkg.go.dev/github.com/bazelbuild/bazel-gazelle/language?tab=doc#GenerateResult) implies they need exactly the same number of items as Resolve will be called with each index in tandem?

(I'm using your PR to get a feel for how gazelle generation works, so I'm new to this and might not be following correctly)

Reply:
If you're looking for a better example, you can check out https://github.com/bazelbuild/bazel-skylib/pull/251

Reply:
Ok, seeing the `setPrivateAttr(config.GazelleImportsKey)` part makes me realize I have no idea why `GenerateResult` has `Gen` and `Imports`...

Sorry for the spam, like I said just wrapping my head around this.

Reply:
> If you're looking for a better example, you can check out [bazelbuild/bazel-skylib#251](https://github.com/bazelbuild/bazel-skylib/pull/251)

ty, I'll take any example I can get 😄